docs/source/clickhouse/schema_design.rst

```diff
@@ -0,0 +1,46 @@
+ClickHouse Schema Design Best Practices
+=======================================
+
+Columns based on dictionary data
+--------------------------------
+
+ClickHouse is a columnar datastore, and at run-time it loads columns on-demand
+based on the columns referenced in the query (both the columns ``SELECT`` ed
+and those part of the ``WHERE`` clause). The ability to store different columns independently
+and not load them for every row for every query is part of the performance advantage that
+ClickHouse provides over a traditional RDBMS (like PostgreSQL).
+
+Commonly, a data schema contains a flexible key:value pair mapping
+(canonically: ``tags`` or ``contexts``) and stores that
+data in a column that contains two ``Nested`` arrays where the first array contains the keys
+of the dictionary and the second array contains the values. This works well when
+your dataset design gives you the ability to filter based on other attributes to a small
+number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
+Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
+specific ``key=value`` pair. In this case one is often best-served by creating a top-level
+column and "promoting" the data to it [#dupe]_.
+
+This promotion is preferable on a couple dimensions:
+
+1. Even though it looks like a dictionary, our tags and contexts "maps" are actually arrays of keys and values.
```

Comment:
This is not 100% correct as it is missing the bloomfilter piece. 
tag keys/values have a bloomfilter index associated. 
It is true that, in the worst case scenario, the complexity of scanning the tag column to find a row with the ``tag=value`` condition is `O(m * n)` but it is also true that such scenario is supposed not to happen in most scenarios as the bloomfilter drops the majority of blocks that do not contain ``tag=value`` before scanning. 

The actual reason to take into account is the selectivity of your tag. That is what makes us decide whether to promote it or not.
Since the bloomfilter index just drops blocks before scanning, we evaluate if it would drop enough blocks.
- When checking if a tag key exists, the bloomfitler would not drop anything if the tag key exists (with different values) on all rows. Think about something like a `request_id` tag. It will be there always.
- When checking ``tag=value``, a value that is low selectivity (present in most rows) would make the bloomfilter discard almost no blocks. Think about an `http_response` tag, the value `200` will be there almost always. The bloomfilter is useless. 

We promote the tag when the tag is there very often, the bloomfilter is not capable of dropping anything and the column is queried very often.